File: msg_passing_algorithm.py
##########################
# Initialize activations
##########################
import numpy as np

neuron_shape = (40, 40)
side_length = neuron_shape[0]
bs = 2
re = 3
ri = 5
sigmaE = 3
we = 30
wi = 5
leaky = we + wi

a = 0.007 * np.random.rand(bs, neuron_shape[0]*neuron_shape[1])
a = np.append(a, [[0],[0]], axis = 1)

print("a = " +str(a))

#####################################
# Compute num_E_nbs and num_I_nbs
#####################################

def get_num_nbs(r):
    count = 0
    for i in range((r+1)**2):
        xi = i // (r+1)
        yi = i % (r+1)
        distsq = xi**2 + yi**2
        if distsq <= r**2:
            count+=1

    num_nbs = (count-(r+1))*4 +1
    return num_nbs

num_E_nbs = get_num_nbs(re)  # 29

num_I_nbs = get_num_nbs(ri)  # 81


print("num_E_nbs = " + str(num_E_nbs))
#print("num_I_nbs = " +str(num_I_nbs))

#####################################
# compute index set of each neurons
#####################################

def get_struct(r):
    rsq = r**2
    l = np.zeros(2*r+1)
    for i in range(2*r+1):
        count = 0
        for j in range(2*r+1):
            distsq = (i - r) ** 2 + (j - r) ** 2
            if distsq <= rsq:
                count += 1
        l[i] = count
    return l

def get_index_from_position(xi, yi, xj, yj, r):
    r = int(r)
    xi = int(xi)
    yi = int(yi)
    xj = int(xj)
    yj = int(yj)
    l = get_struct(r)
    core_index = (get_num_nbs(r)-1)/2
    if yi == yj:
        index = core_index + (xj - xi)
        index = int(index)
        return index
    else:
        diff = 0
        for i in range(int(abs(yj - yi)-1)):
            diff += l[r-i-1]
        if (yi > yj):
            index = core_index - (l[r]-1)/2 - diff - (l[r-(yi - yj)]+1)/2 + (xj - xi)
            index = int(index)
            return index
        elif (yi < yj):
            index = core_index + (l[r]-1)/2 + diff + (l[r-(yi - yj)]+1)/2 + (xj - xi)
            index = int(index)
            return index


def compute_indexset(r, num_nbs):
    set = np.zeros(shape=(neuron_shape[0]*neuron_shape[1], num_nbs))
    set = set.astype(int)
    set.fill(neuron_shape[0]*neuron_shape[1])
    #print(set)
    for i in range(neuron_shape[0]*neuron_shape[1]):
        xi = i // side_length
        yi = i % side_length
        for j in range(neuron_shape[0]*neuron_shape[1]):
            xj = j//side_length
            yj = j % side_length
            distsq = (xi - xj)**2 + (yi - yj)**2
            if distsq <= r**2:
                index = get_index_from_position(xi, yi, xj, yj, r)
                index = int(index)
                #print("index = " + str(index))
                set[i][int(index)] = j
                #print(set[i][index])
    return set

N_E = compute_indexset(re, num_E_nbs)
N_I = compute_indexset(ri, num_I_nbs)

#print("N_E = " +str(N_E))
#print("N_I = " +str(N_I))

#####################################
# compute weight kernels W_E and W_I
#####################################
W_E = np.zeros(num_E_nbs)
W_I = np.zeros(num_I_nbs)

count_E = 0
for i in range((2*re+1)**2):
    xi = i // (2*re+1)
    yi = i % (2*re+1)
    distsq = (xi - re)**2 + (yi - re)**2
    if distsq <= re**2:
        W_E[count_E] = np.exp(- distsq/2/sigmaE)
        count_E += 1

#print("count_E = " + str(count_E))
W_E = we * W_E / np.sum(W_E)
print("W_E = " +str(W_E))

count_I = 0
for i in range((2*ri+1)**2):
    xi = i // (2 * ri + 1)
    yi = i % (2 * ri + 1)
    distsq = (xi - ri) ** 2 + (yi - ri) ** 2
    if distsq <= ri**2:
        W_I[count_I] = 1
        count_I += 1

#print("count_I = " + str(count_I))
W_I = wi * W_I / np.sum(W_I)
#print("W_I = " +str(W_I))

###########################
# Update algorithms
###########################

def activation_update(a):
    b = np.zeros(shape=(bs, neuron_shape[0] * neuron_shape[1]))
    for k in range(bs):
        r = np.zeros(neuron_shape[0]*neuron_shape[1])
        for i in range(neuron_shape[0]*neuron_shape[1]):
            r[i] = - leaky * a[k][i]
            for j in range(num_E_nbs):
                r[i] += W_E[j] * a[k][N_E[i][j]]
            for j in range(num_I_nbs):
                r[i] -= W_I[j] * a[k][N_I[i][j]]

        for i in range(neuron_shape[0]*neuron_shape[1]):
            a[k][i] = r[i]

    for k in range(bs):
        for n in range(neuron_shape[0]*neuron_shape[1]):
            b[k][n] = a[k][n]
    return b


b = activation_update(a)

print("b = " +str(b))

File: c_version/msg_passing_algorithm.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include "matrix_float.c"

void print_int_matrix(int row, int col, int** m) {
    for (int i = 0; i < row; i++) {
        for (int j = 0; j < col; j++) {
            printf("%d ", m[i][j]);
        }
        printf("\n");
    }
}

int get_num_nbs(int r){
    int count = 0;
    for (int i = 0; i < (r+1)*(r+1); i++){
        int xi = i/(r+1);
        int yi = i%(r+1);
        int distsq = xi * xi + yi * yi;
        if (distsq <= r*r ){
            count += 1;
        }
    }

    int num_nbs = (count - (r + 1)) * 4 + 1;
    return num_nbs;
}

int* get_struc(int r){
    int rsq = r*r;
    int* l = malloc(sizeof(int) * (2*r + 1));
    for (int i = 0; i < 2*r+1; i++){
        int count  = 0;
        for (int j =0; j < 2*r+1; j++){
            int distsq = (i - r) * (i - r) + (j - r) * (j - r);
            if (distsq <= rsq){
                count+=1;
            }
        }
        l[i] = count;
    }
    return l;
}

int get_index_from_position(int xi, int yi, int xj, int yj, int r){
    int* l = get_struc(r);
    int core_index = (get_num_nbs(r) - 1)/2;
    if (yi == yj){
        int index = core_index + (xj - xi);
        free(l);
        return index;
    }
    else{
        int diff = 0;
        for (int i = 0; i < abs(yj - yi)-1; i++){
            diff += l[r-i-1];
        }
        if (yi > yj){
            int index = core_index - (l[r]-1)/2 - diff - (l[r-(yi - yj)]+1)/2 + (xj - xi);
            free(l);
            return index;
        }
        else {
            int index = core_index + (l[r]-1)/2 + diff + (l[r-(yi - yj)]+1)/2 + (xj - xi);
            free(l);
            return index;
        }
    }
}

int ** compute_indexset(r, num_nbs, neuron_shape){
    int side_length = sqrt(neuron_shape);
    int** set = malloc(sizeof(int*) * neuron_shape);
    for (int i = 0; i < neuron_shape; i++) {
        set[i] = malloc(sizeof(float) * num_nbs);
    }
    for (int i = 0; i < neuron_shape; i++){
        for (int j = 0 ; j < num_nbs; j++){
            set[i][j] = neuron_shape;
        }
    }
    for (int i = 0; i < neuron_shape; i++){
        int xi = i / side_length;
        int yi = i % side_length;
        for (int j = 0; j < neuron_shape; j++){
            int xj = j / side_length;
            int yj = j % side_length;
            int distsq = (xi - xj)*(xi - xj) + (yi - yj)*(yi - yj);
            if (distsq <= r*r){
                int index = get_index_from_position(xi, yi, xj, yj, r);
                set[i][index] = j;
            }
        }
    }

    return set;
}

float* compute_WE(int num_nbs, int r, int we, int sigmaE){
    float* W = malloc(sizeof(float) * num_nbs);
    int count = 0;
    for (int i = 0; i < (2*r+1)*(2*r+1); i++){
        int xi = i / (2*r + 1);
        int yi = i % (2*r + 1);
        int distsq = (xi - r)* (xi - r) + (yi - r)* (yi - r);
        if (distsq <= r*r){
            W[count] = exp(- distsq/2.0/sigmaE);
            count += 1;
        }
    }

    float W_E_sum = sum(num_nbs, W);
    for (int i = 0; i < num_nbs; i++){
        W[i] = we * W[i] / W_E_sum;
    }

    return W;
}

float* compute_WI(int num_nbs, int r, int wi){
    float* W = malloc(sizeof(float) * num_nbs);
    int count = 0;
    for (int i = 0; i < (2*r+1)*(2*r+1); i++){
        int xi = i / (2*r + 1);
        int yi = i % (2*r + 1);
        int distsq = (xi - r)* (xi - r) + (yi - r)* (yi - r);
        if (distsq <= r*r){
            W[count] = 1;
            count += 1;
        }
    }

    float W_E_sum = sum(num_nbs, W);
    for (int i = 0; i < num_nbs; i++){
        W[i] = wi * W[i] / W_E_sum;
    }

    return W;
}

float** exc_act_update(float** exc_act_dummy, float** inh_act_dummy, int bs, int neuron_shape, int leaky,
                       int num_E_nbs, int num_I_nbs, float* W_E, float* W_I, int** N_E, int** N_I){
    float** b = malloc_matrix(bs, neuron_shape);
    for (int k = 0; k < bs; k++){
        float* r = malloc(sizeof(float) * neuron_shape);
        for (int i = 0; i < neuron_shape; i++){
            r[i] = - leaky * exc_act_dummy[k][i];
            for (int j = 0; j < num_E_nbs; j++){
                r[i] += W_E[j] * exc_act_dummy[k][N_E[i][j]];}
            for (int j = 0; j < num_I_nbs; j++){
                r[i] -= W_I[j] * inh_act_dummy[k][N_I[i][j]];}

            b[k][i] = r[i];
        }
        free(r);
    }
    return b;
}

float** inh_act_update(float** exc_act_dummy, float** inh_act_dummy, int bs, int neuron_shape, int leaky,
                       int num_E_nbs, float* W_E, int** N_E){
    float** b = malloc_matrix(bs, neuron_shape);
    for (int k = 0; k < bs; k++){
        float* r = malloc(sizeof(float) * neuron_shape);
        for (int i = 0; i < neuron_shape; i++){
            r[i] = - leaky * inh_act_dummy[k][i];
            for (int j = 0; j < num_E_nbs; j++){
                r[i] += W_E[j] * exc_act_dummy[k][N_E[i][j]];}

            b[k][i] = r[i];
        }
        free(r);
    }
    return b;
}


int main(){

    int neuron_shape = 1600;
    int side_length = sqrt(neuron_shape);
    int bs = 2;
    int re = 3;
    int ri = 5;
    int sigmaE = 3;
    int we = 30;
    int wi = 5;
    int leaky = we + wi;

    int num_E_nbs = get_num_nbs(re);

    int num_I_nbs = get_num_nbs(ri);

//    printf("num_E_nbs = %d\n", num_E_nbs);
//    printf("num_I_nbs = %d\n", num_I_nbs);
//    int r = 3;
//    int* l = get_struc(r);
//    print_int_vector(2*r+1, l);

//    int r = 5;
//    int index = get_index_from_position(7, 5, 8, 8, r);
//    printf("index = %d\n", index);

    int** N_E = compute_indexset(re, num_E_nbs, neuron_shape);
    int** N_I = compute_indexset(ri, num_I_nbs, neuron_shape);

//    print_int_matrix(neuron_shape, num_E_nbs, N_E);
//    printf("\n");
//    print_int_matrix(neuron_shape, num_I_nbs, N_I);
//    printf("\n");

    float* W_E = compute_WE(num_E_nbs, re, we, sigmaE);
    print_float_vector(num_E_nbs, W_E);
    printf("\n");

    float* W_I = compute_WI(num_I_nbs, ri, wi);
    print_float_vector(num_I_nbs, W_I);

}
